src/component-manager/panel.ts

```diff
@@ -0,0 +1,143 @@
+ * File Created: Wednesday, 6th January 2021 1:25:19 pm
+ * Copyright 2021 Espressif Systems (Shanghai) CO LTD
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { join } from "path";
+import {
+  Disposable,
+  Uri,
+  ViewColumn,
+  Webview,
+  WebviewPanel,
+  window,
+} from "vscode";
+import { ESP } from "../config";
+import { readParameter } from "../idfConfiguration";
+import { Logger } from "../logger/logger";
+
+export class ComponentManagerUIPanel {
+  private static instance: ComponentManagerUIPanel;
+
+  private readonly panel: WebviewPanel;
+  private disposable: Disposable[] = [];
+  private readonly extensionPath: string;
+
+  public static show(extensionPath: string) {
+    const column = window.activeTextEditor
+      ? window.activeTextEditor.viewColumn
+      : undefined;
+    if (!!ComponentManagerUIPanel.instance) {
+      ComponentManagerUIPanel.instance.panel.reveal(column);
+    }
+    const url = readParameter("esp.component-manager.url");
+    if (!url) {
+      throw new Error("esp.component-manager.url is not set");
+    }
+    const panel = window.createWebviewPanel(
+      ESP.Webview.ComponentManagerUI.ViewType,
+      ESP.Webview.ComponentManagerUI.Title,
+      column || ViewColumn.One,
+      {
+        enableScripts: true,
+        retainContextWhenHidden: true,
+        enableCommandUris: true,
+      }
+    );
+    ComponentManagerUIPanel.instance = new ComponentManagerUIPanel(
+      panel,
+      extensionPath,
+      url
+    );
+  }
+
+  public dispose() {
+    ComponentManagerUIPanel.instance = null;
+    this.panel.dispose();
+    while (this.disposable.length) {
+      const d = this.disposable.pop();
+      d ? d.dispose() : null;
+    }
+  }
+
+  private constructor(panel: WebviewPanel, extensionPath: string, url: string) {
+    this.panel = panel;
+    this.extensionPath = extensionPath;
+    this.panel.onDidDispose(() => this.dispose(), null, this.disposable);
+    this.panel.webview.onDidReceiveMessage(
+      (e) => this.onMessage(e),
+      null,
+      this.disposable
+    );
+    this.panel.iconPath = Uri.file(
+      join(extensionPath, "media", "espressif_icon.png")
+    );
+    this.panel.webview.html = this.initWebView(url);
+  }
+  private onMessage(message: any) {
+    switch (message.message) {
+      case "install":
+        if (message.url) {
+          Logger.infoNotify(`Installing Component from ${message.url}`);
+        }
```

Comment:
@kumekay what's the command we need to run to install this component?

Reply:
@pwmb There is no such command in component manager since real dependency solving is only possible during run of CMake, 
what extension can do is to add a dependency to `idf_component.yml` file. This task was not implemented yet in component manager CLI, because there are no well-maintained yaml library for python with support for comments in yaml and we cannot reliably modify user-maintained yaml files.
So you can either implement it in extension itself, or I can add this feature.

If you want to actually download components after adding them to manifest you have to rerun CMake, however I think we shouldn't do it by default, because it may be slow.

Reply:
I think it will be better if we have a sub-command (in `idf.py` or relevant tool) to add entry to the manifest file and when user will build the project the dependency will be downloaded.

Reply:
Sure, it will be a part of `idf.py`,
Since it's possible to add dependencies not only to the main component of the project, but also for own components in `components` directory, full command will look like:

`idf.py add-dependency --component=main  example/cmp>=2.0.0 --add=expample/cmp2==1.0.0`

Reply:
We can only support adding dependencies to `main` component for projects or to components itself, if currently open code is idf component

Reply:
Going to work on the feature this week

Reply:
@brianignacio5 we can inject the command which is needed to add the deps once its available from IDF side, for now we are exposing the download url from the Component Manager UI, but I think we can expose more entities as required

Reply:
Sure, I'm sorry we changed the focus for last to sprints, so this feature is postponed a bit.